lib/navigation/preprocess_encoders.cpp

```diff
@@ -16,41 +31,101 @@ core::EncoderData EncodersPreprocessor::processData(const core::EncoderData)
-  TODOLater: implement
-  rough process:
-  - get q1, median, q3 of encoder array
-  - define upper & lower bounds as (-)1.5*inter-quatrile range
-  - if any datapoint outwith this, set
-  num_outliers_per_encoder_[i] += 1
-  -set outlier points as median
-
-  -also has to be able to handle 1 unreliable sensor
-  -also figure out return type/ what we update and update
-  documentation as appropriate
-  */
-  return {0, 0, 0, 0};
+  core::EncoderData encoder_data_copy;
+  std::copy(encoder_data.begin(), encoder_data.end(), encoder_data_copy.begin());
+  Quartile quartiles;
+  core::Float interquartile_range; // no longer const
+  core::Float upper_bound;
+  core::Float lower_bound;
+  if(num_reliable_encoders_ == core::kNumEncoders) {
+    core::EncoderData reliable_data;
+    std::copy(
+      encoder_data.begin(), encoder_data.end(), reliable_data.begin());
+    quartiles = getQuartiles(reliable_data); 
+    interquartile_range        = quartiles.q3 - quartiles.q1;
+    upper_bound = quartiles.median + 1.5 * interquartile_range ;
+    lower_bound = quartiles.median - 1.5 * interquartile_range ;
+
+  }
+  else if (num_reliable_encoders_ == core::kNumEncoders - 1) {
+    std::array<uint32_t, core::kNumEncoders - 1> reliable_data;
+    std::size_t counter = 0;
+    for (std::size_t i = 0; i < encoder_data.size(); ++i) {
+      if (encoders_reliable_.at(i)) {
+        reliable_data.at(counter) = encoder_data.at(i);
+        ++counter;
+      }
+    }
+    quartiles = getQuartiles(reliable_data);
+    interquartile_range        = quartiles.q3 - quartiles.q1;
+    upper_bound = quartiles.median + 1.2 * interquartile_range ;
+    lower_bound = quartiles.median - 1.2 * interquartile_range ;
+  }
+  else{
+    logger_.log(core::LogLevel::kFatal, "Sensors have become unreliable");
+    return std::nullopt;
+  }
+
+  for (std::size_t i = 0; i < encoder_data_copy.size(); ++i) {
+    if (encoder_data_copy.at(i) > upper_bound || encoder_data_copy.at(i) < lower_bound || encoders_reliable_.at(i) == false) {
+      encoder_data_copy.at(i)         = quartiles.median;
+
+      if(encoders_reliable_.at(i)){
+        encoder_outliers_.at(i) = encoder_outliers_.at(i) + 1;
+      }
+    } else {
+      encoder_outliers_.at(i) = 0;
+    }
+  }
+
+  return encoder_data_copy;
 }
 
-void EncodersPreprocessor::checkReliable(const core::EncoderData &encoder_data)
+
+
+SensorChecks EncodersPreprocessor::checkReliable()
 {
-  /*
-  TODOLater: implement
-  rough process:
-  - check how many times an individual encoder
-  has been an outlier in a row (outlier encoders)
-  -if num_outliers_per_encoder_[i] > n (tbd), mark encoder as unreliable
-  in reliable encoders.
-  -if number unreliable in reliable_encoders > 1, fail state
-
-  -also figure out return type/ what we update and update
-  documentation as appropriate
-  */
+  for (std::size_t i = 0; i < core::kNumEncoders; ++i) {
+    if (encoder_outliers_.at(i) > max_consecutive_outliers_) {  // for now assuming n to be 10
```

Comment:
You need to turn this into an and statement to also check if it has already been classified as unreliable. Otherwise you will set an unreliable sensor as still false but then still subtract 1 from the number of reliable imus which will cause critical failure every time a single sensor fails.